easy_logging.go

```diff
@@ -0,0 +1,138 @@
+package gosnowflake
+
+import (
+	"errors"
+	"io"
+	"os"
+	"path"
+	"strings"
+)
+
+type initTrials struct {
+	everTriedToInitialize bool
+	initParameter         string
+	configureCounter      int
+}
+
+var easyLoggingInitTrials = initTrials{
+	everTriedToInitialize: false,
+	initParameter:         "",
+	configureCounter:      0,
+}
+
+func (i *initTrials) setInitTrial(initParameter string) {
+	i.everTriedToInitialize = true
+	i.initParameter = initParameter
+}
+
+func (i *initTrials) increaseReconfigureCounter() {
+	i.configureCounter++
+}
+
+func (i *initTrials) resetInitTrials() {
+	i.everTriedToInitialize = false
+	i.initParameter = ""
+	i.configureCounter = 0
+}
+
+func initEasyLogging(configFilePathFromConnectionString string) error {
+	if !allowedToInitialize(configFilePathFromConnectionString) {
+		return nil
+	}
+	config, err := getClientConfig(configFilePathFromConnectionString)
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		easyLoggingInitTrials.setInitTrial(configFilePathFromConnectionString)
+		return nil
+	}
+	var logLevel string
+	logLevel, err = getLogLevel(config.Common.LogLevel)
+	if err != nil {
+		return err
+	}
+	var logPath string
+	logPath, err = getLogPath(config.Common.LogPath)
+	if err != nil {
+		return err
+	}
+	reconfigureEasyLogging(logLevel, logPath)
+	easyLoggingInitTrials.setInitTrial(configFilePathFromConnectionString)
+	easyLoggingInitTrials.increaseReconfigureCounter()
+	return nil
+}
+
+func reconfigureEasyLogging(logLevel string, logPath string) error {
+	err := logger.SetLogLevel(logLevel)
+	if err != nil {
+		return err
+	}
+	var output io.Writer
+	output, err = createLogWriter(logPath)
+	if err != nil {
+		return err
+	}
+	logger.SetOutput(output)
+	return nil
+}
+
+func createLogWriter(logPath string) (io.Writer, error) {
+	if strings.EqualFold(logPath, "STDOUT") {
+		return os.Stdout, nil
+	}
+	logFileName := path.Join(logPath, "snowflake.log")
+	file, err := os.OpenFile(logFileName, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+	if err != nil {
+		return nil, err
+	}
+	return io.MultiWriter(file, os.Stdout), nil
+}
+
+func allowedToInitialize(configFilePathFromConnectionString string) bool {
+	triedToInitializeWithoutConfigFile := easyLoggingInitTrials.everTriedToInitialize && easyLoggingInitTrials.initParameter == ""
```

Comment:
I'm not sure what this `initParameter` means, can you explain it a bit?

Reply:
it is Config.ClientConfigFile so the parameter the user provided in the config or dsn

Reply:
Please, rename it to be more descriptive.

Reply:
renamed to: clientConfigFileInput